library/java/io/envoyproxy/envoymobile/Envoy.java

```diff
@@ -9,22 +9,67 @@
 // Wrapper class that allows for easy calling of Envoy's JNI interface in native Java.
 public class Envoy {
 
-  public void load() { System.loadLibrary("envoy_jni"); }
+  // Internal reference to helper object used to load and initialize the native library.
+  // Volatile to ensure double-checked locking works correctly.
+  private static volatile EnvoyLoader loader = null;
 
-  public void run(Context context, String config) {
-    Thread thread = new Thread(new Runnable() {
+  // Dedicated thread for running this instance of Envoy.
+  private final Thread runner;
+
+  // Private helper class used by the load method to ensure the native library and its
+  // dependencies are loaded and initialized at most once.
+  private static class EnvoyLoader {
```

Comment:
If/when we translate this to Kotlin we can avoid nesting this class within `Envoy` :)

Reply:
cool